src/pip/_internal/operations/prepare.py

```diff
@@ -223,6 +223,7 @@ def unpack_url(
     download,  # type: Downloader
     download_dir=None,  # type: Optional[str]
     hashes=None,  # type: Optional[Hashes]
+    in_tree_build=False,  # type: bool
```

Comment:
It feels a bit weird to have this argument to unpack_url, because an unpack operation should not be concerned about building. Maybe the caller should not invoke unpack_url at all if `link.is_existing_dir and in_tree_build`. But maybe it's ok to have it as it will be removed at the end of the deprecation period and it signals we have cleanup to do here at that moment.

Reply:
We usually notice deprecations late in the release process -- I'd prefer we keep the deprecation-removal step as small as possible, performing any possible reasonable cleanup beforehand.

I'm also A-OK with adding a comment above the deprecated call, to note any details for how to actually do the deprecation.

Reply:
@pradyunsg one of the show stoppers that led to the revert of #7882 was https://github.com/pypa/pip/issues/8165#issuecomment-624669107 (the change silently producing invalid build artifacts).

So my worry is that people will not notice the warning and wrong builds will happen silently when we flip the default.